### PR TITLE
feat(nx): update to NgRx 8 and add schematics support for creators and entities

### PR DESCRIPTION
**Pull request by brandonroberts**
- Bumps NgRx dependencies to latest version
- Bumps jasmine-marbles dependency to `0.6.0`
- Adds a devDependency on `@ngrx/schematics` when setting up NgRx
- Adds support for NgRx creator functions in schematics and DataPersistence service
- Uses `@ngrx/entity` by default for managing collections
- Also porting setup changes upstream for less maintenance
- Updated NgRx guide and API docs

### CODE REVIEW

**Comment by brandonroberts** [comment]
Still working on adding a migration to `@nrwl/angular` for breaking changes in NgRx 8. Hoping to use existing upstream migrations

**Comment by wesleygrimes** [comment]
I'll take sometime and review this today.

File: docs/angular/guides/misc-ngrx.md [hunk 1]
```diff
@@ -1,484 +1,119 @@
...
-
-The `ngrx` schematic generates a NgRx feature set containing the following files:
+Nx provides a schematic to build out a new NgRx feature area that manages shared state.
 
-- `actions`,
-- `reducer`,
-- `effects`,
-- `selectors`, and
-- `facade` (optional)
+The **@nrwl/angular** package has an `ngrx` schematic to generate files that implement best practices when using NgRx for state management. This schematic generates source files, along with enhancements the generated NgRx for data persistence strategies, and simplified testing.
```
Comment (wesleygrimes) [inline docs/angular/guides/misc-ngrx.md]:
```suggestion
The **@nrwl/angular** package has an `ngrx` schematic to generate files that implement best practices when using NgRx for state management. This schematic generates source files that include enhancements to NgRx for data persistence strategies, and simplified testing.
```

File: docs/angular/guides/misc-ngrx.md [hunk 2]
```diff
@@ -1,484 +1,119 @@
...
 ---
 
-### Root
-
-Making use of the Angular CLI ng generate command, we can use the ngrx schematic to scaffold out the following in our applications:
+## Initial Setup
 
-- Root level NgRx configuration
-- Feature level NgRx configuration
+To get started using with NgRx in an Angular application, you set up the root level store. As your application grows, you add feature level states, ensuring that your code follows a common pattern each time.
```
Comment (wesleygrimes) [inline docs/angular/guides/misc-ngrx.md]:
```suggestion
To get started with NgRx in an Angular application, you set up the root level store. As your application grows, you add feature level states, ensuring that your code follows a common pattern each time.
```

File: docs/angular/guides/misc-ngrx.md [hunk 3]
```diff
@@ -1,484 +1,119 @@
...
+ng g @nrwl/angular:ngrx products --module=libs/products/src/lib/products.module.ts --directory +state/products --syntax=creators --defaults
 ```
 
-> This would use `StoreModule.forRoot()` to register the Comments NgRx state functionality.
+> Use the `--facade` option to generate an injectable Facade class along with the feature.
 
-<br/>
+The following files are created, or updated:
 
-Better yet, let's generate a `Comments` feature set within a `state` library and register it with the `comments-state.module.ts` file in the same `comments/state` folder.
```
Comment (wesleygrimes) [inline docs/angular/guides/misc-ngrx.md]: Looks like we decided to remove most of the content? Was it too wordy? I can see where value is added so that the reader can see exactly the type of content being output when these files are generated.

File: docs/angular/guides/misc-ngrx.md [hunk 4]
```diff
@@ -1,484 +1,119 @@
...
-## Learn by Example
+To manage the feature state:
 
-Consider a command to generate a `Comments` NgRx feature set and register it within an application root ngModule.
+- Use the `ngrx` schematic with the feature name in plural form, such as `products`.
+- Provide a path to the `products` library module.
 
 ```bash
-ng g @nrwl/angular:ngrx Comments --root --module=apps/myapp/src/app/app.module.ts
+ng g @nrwl/angular:ngrx products --module=libs/products/src/lib/products.module.ts --directory +state/products --syntax=creators --defaults
```
Comment (wesleygrimes) [inline docs/angular/guides/misc-ngrx.md]: Below in the API docs for the schematics options it shows "Products" in title case, but here it shows lowercase. Which is correct?

File: docs/api-angular/schematics/ngrx.md [hunk 5]
```diff
@@ -17,64 +17,88 @@ Default: `+state`
...
 
-Path to ngModule; host directory will contain the new '+state' directory (e.g., libs/comments/src/lib/comments-state.module.ts).
+The path to NgModule where the feature state will be registered. The host directory will create/use the new state directory.
 
 ### name
 
 Type: `string`
 
-Name of the NgRx feature (e.g., Products, Users, etc.). Recommended to use plural form for name.
+Name of the NgRx feature state, such as "Products" or "Users"). Recommended to use the plural form of the name.
```
Comment (wesleygrimes) [inline docs/api-angular/schematics/ngrx.md]: Should the feature state name be title case or all lower case, or does it even matter?

File: packages/angular/src/schematics/ngrx/ngrx.spec.ts [hunk 6]
```diff
@@ -207,10 +230,38 @@ describe('ngrx', () => {
 
     const appModule = getFileContent(tree, '/apps/myapp/src/app/app.module.ts');
     expect(appModule).not.toContain('StoreModule');
-    expect(appModule).not.toContain(
-      '!environment.production ? [storeFreeze] : []'
+    expect(appModule).not.toContain('!environment.production ? [] : []');
+
+    [
```
Comment (wesleygrimes) [inline packages/angular/src/schematics/ngrx/ngrx.spec.ts]: Should we also test for the existence of `state.models.ts` if the `creators` option is chosen?

File: packages/angular/src/schematics/ngrx/schema.json [hunk 7]
```diff
@@ -1,59 +1,75 @@
...
   "$schema": "http://json-schema.org/schema",
   "id": "SchematicsNxNgrx",
-  "title": "Add NgRx support to a module",
+  "title": "Add NgRx support to an application or library",
   "type": "object",
   "properties": {
     "name": {
       "type": "string",
-      "description": "Name of the NgRx feature (e.g., Products, Users, etc.). Recommended to use plural form for name.",
+      "description": "Name of the NgRx feature state, such as \"Products\" or \"Users\"). Recommended to use the plural form of the name.",
```
Comment (wesleygrimes) [inline packages/angular/src/schematics/ngrx/schema.json]: As previously mentioned, should `Products` be title case or all lower case or does it even matter as the schematic will convert to lower?

File: packages/angular/src/utils/versions.ts [hunk 8]
```diff
@@ -2,7 +2,7 @@ export const nxVersion = '*';
 export const angularVersion = '^8.0.0';
 export const angularDevkitVersion = '^0.800.0';
 export const angularJsVersion = '1.6.6';
-export const ngrxVersion = '7.2.0';
+export const ngrxVersion = '8.1.0';
 export const ngrxStoreFreezeVersion = '0.2.4';
```
Comment (wesleygrimes) [inline packages/angular/src/utils/versions.ts]:
```suggestion
```

I believe we can remove this now as it is no longer used anywhere.

**Review by wesleygrimes — CHANGES_REQUESTED** [review]
Looks good Brandon! Just a few nits.

File: docs/angular/guides/misc-ngrx.md (hunk 4 above)
Comment (brandonroberts) [reply docs/angular/guides/misc-ngrx.md]: I've changed the references to lowercase "products" for clarity. It works either way

File: docs/angular/guides/misc-ngrx.md (hunk 3 above)
Comment (brandonroberts) [reply docs/angular/guides/misc-ngrx.md]: I think its too wordy to have all this output here. If there was a better way to show/hide file contents I think it would be better.

File: packages/angular/src/schematics/ngrx/rules/add-imports-to-module.ts [hunk 9]
```diff
@@ -25,105 +33,136 @@ export function addImportsToModule(context: RequestContext): Rule {
...
+      module: relativeModuleName,
+      path: moduleDir,
+      root: true
+    });
+
+    /**
+     * Delegate the effects setup to NgRx schematics
+     */
+
+    const setupRootEffects = externalSchematic('@ngrx/schematics', 'effect', {
```
Comment (FrozenPandaz) [inline packages/angular/src/schematics/ngrx/rules/add-imports-to-module.ts]: Setting up root Effects seems to cause `app.module.ts` to become malformed:

image

File: packages/angular/src/utils/versions.ts (hunk 8 above)
Comment (FrozenPandaz) [reply packages/angular/src/utils/versions.ts]: Yes, it can be removed :)

**Review by FrozenPandaz — CHANGES_REQUESTED** [review]
Awesome! :rocket: 

The ngrx E2E Tests are failing.

I think the reason is delegating to `@ngrx/schematics` makes the user fail if they don't have `@ngrx/schematics` installed.. and also for generating root, they need `@ngrx/router-store` installed?

Possible options I see are...
* Don't delegate :thinking:
* Delegate but require the user to install them :-1:
* Delegate but install it as a dependency of `@nrwl/angular` :woman_shrugging:

File: packages/angular/src/schematics/ngrx/rules/add-imports-to-module.ts (hunk 9 above)
Comment (brandonroberts) [reply packages/angular/src/schematics/ngrx/rules/add-imports-to-module.ts]: The AppFacade thing is weird. I'll look into it
The `.forRoot()` and `.forFeature()` is intentional. This is to satisfy the existing use case of registering the root with a feature. In general, we want to encourage developers to keep the root object empty, and use `forFeature()` to register additional feature states.

**Comment by brandonroberts** [comment]
@FrozenPandaz thanks for review. Is the installation of the packages not guaranteed if we modify the `package.json` and add an install task before the schematics are run?

**Comment by brandonroberts** [comment]
I reverted the delegation to `@ngrx/schematics`, and updated the imports to follow the same general practices for setup. I also added the `--routing` option to the test to make sure it continues functioning during future changes.

**Comment by FrozenPandaz** [comment]
Migration is broken for unrelated reasons but this looks good.

**Review by FrozenPandaz — APPROVED** [review]

File: packages/angular/src/migrations/update-8-3-0/upgrade-ngrx-8-0.ts [hunk 10]
```diff
@@ -0,0 +1,24 @@
+import {
+  chain,
+  Tree,
+  SchematicContext,
+  noop
+} from '@angular-devkit/schematics';
+import { addUpdateTask, readJsonInTree, formatFiles } from '@nrwl/workspace';
+import { ngrxVersion } from '../../utils/versions';
+
+const runNgrxUpdate = addUpdateTask('@ngrx/store', ngrxVersion);
```
Comment (FrozenPandaz) [inline packages/angular/src/migrations/update-8-3-0/upgrade-ngrx-8-0.ts]: Let's use a hard coded version here in case people are migrating to not the latest version.

**Review by FrozenPandaz — APPROVED** [review]

File: packages/angular/src/schematics/ngrx/ngrx.spec.ts (hunk 6 above)
Comment (wesleygrimes) [reply packages/angular/src/schematics/ngrx/ngrx.spec.ts]: @brandonroberts What are your thoughts about adding a bit to test for `state.models.ts` as mentioned above?

**Comment by wesleygrimes** [comment]
Just had one outstanding question about adding to `ngrx.spec.ts` for `state.models.ts` if `creators` if chosen. Other than that I am good.

**Comment by brandonroberts** [comment]
I have a test for the models file here https://github.com/nrwl/nx/pull/1550/files#diff-9cf1f28d0209d3021f9d79dfca3b67f2R626

**Comment by wesleygrimes** [comment]
Perfect. I’m good with this.

**Review by wesleygrimes — APPROVED** [review]
